### PR TITLE
Added `assign-owner-group` command

### LINKED ISSUE #2890: [BUG]: Change in UC Catalog Ownership to Hive Metastore Schema Owner After Executing UCX Table Migration workflow.
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

NFCU created a UC catalog with an SPN as the owner, managed by Terraform. After running the table migration workflow in UCX, the ownership of the UC catalog changes from the SPN to the owner of the recently migrated Hive metastore schema. Since these schemas are owned by human users, the UC catalog ownership is shifting to them. This issue has been replicated multiple times, and screenshots have been collected for reference, raising significant security concerns. They are currently using version 0.38.0 of UCX.

### Expected Behavior

After executing the UCX table migration, the UC catalog's ownership should remain with the Service Principal Name (SPN) that created it rather than the recent Hive Metastore schema owner. 

### Steps To Reproduce

1)Create a catalog using a Service Principal (in the DevOps pipeline), with the SPN as the catalog owner.
2)Set up a Hive metastore schema and table, assigning ownership to a human user (this ownership is visible from a UC-enabled cluster but not from a non-UC-enabled cluster).
3)Execute the UCX assessment workflow.
4)Update the table_mappings.csv file with the schema and table created in step 2.
5)Run the UCX table migration workflow.
6)Observe the ownership of the UC catalog change from the SPN to the owner of the schema created in step 2.

UCX version used: 0.38 

Other details:
1) Customer couldn't **Create Uber principal.** The tables with mount point references were migrated without issue.
2) Table Access control workspace setting is set to "off" 

### Cloud

Azure

### Operating System

Linux

### Version

via install.sh

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by FastLee**
replaces #3075 
close #2890

### CODE REVIEW

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/config.py [hunk 1]
```diff
@@ -78,6 +78,12 @@ class WorkspaceConfig:  # pylint: disable=too-many-instance-attributes
 
     managed_table_external_storage: str = 'CLONE'
 
+    # Skip ACL Migration
+    skip_grant_migration: bool = False
```
[the remark this answers is not in the document]
Comment (FastLee) [reply src/databricks/labs/ucx/config.py:82]: Done

File: src/databricks/labs/ucx/hive_metastore/grants.py [hunk 2]
```diff
@@ -776,14 +776,27 @@ def __init__(
         self,
         sql_backend: SqlBackend,
         group_manager: GroupManager,
+        ownership_loader: Callable[[], Iterable[Grant]],
         grant_loaders: list[Callable[[], Iterable[Grant]]],
+        /,
+        skip_grant_migration: bool = False,
```
[the remark this answers is not in the document]
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/grants.py:782]: Done

File: src/databricks/labs/ucx/hive_metastore/ownership.py [hunk 3]
```diff
@@ -85,6 +86,94 @@ def _grants_snapshot(self):
         return self._grants_crawler.snapshot()
 
 
+class StaticTableOwnership(Ownership[Table]):
```
[the remark this answers is not in the document]
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/ownership.py:89]: Done

File: src/databricks/labs/ucx/hive_metastore/ownership.py [hunk 4]
```diff
@@ -85,6 +86,94 @@ def _grants_snapshot(self):
...
+            return None, table.name
+        return table.name, None
+
+    def _static_owner(self) -> str | None:
+        if self._fixed_owner_group:
+            return self._fixed_owner_group
+        return self._application_principal
+
+    def _maybe_direct_owner(self, record: Table) -> str | None:
+        return self._static_owner()
```
[the remark this answers is not in the document]
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/ownership.py:149]: Done

File: src/databricks/labs/ucx/hive_metastore/grants.py [hunk 5]
```diff
@@ -835,12 +866,38 @@ def _match_grants(self, src: SecurableObject) -> list[Grant]:
...
+
     def _replace_account_group(self, grant: Grant) -> Grant:
         target_principal = self._workspace_to_account_group_names.get(grant.principal)
         if not target_principal:
             return grant
         return replace(grant, principal=target_principal)
 
+    def _apply_ownership(self, dst: SecurableObject, owner: str):
+        if dst.kind == "TABLE":
+            sql = f"ALTER TABLE {escape_sql_identifier(dst.full_name)} " f"SET OWNER TO {escape_sql_identifier(owner)}"
```
[the remark this answers is not in the document]
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/grants.py:883]: Done

File: src/databricks/labs/ucx/hive_metastore/grants.py [hunk 6]
```diff
@@ -798,11 +811,17 @@ def apply(self, src: SecurableObject, dst: SecurableObject) -> bool:
                 logger.warning(
                     f"failed-to-migrate: Failed to migrate ACL for {src.full_name} to {dst.full_name}", exc_info=e
                 )
-        return True
+        return False
 
     def retrieve(self, src: SecurableObject, dst: SecurableObject) -> list[Grant]:
         grants = []
-        for grant in self._match_grants(src):
+        discover_grants = []
```
[the remark this answers is not in the document]
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/grants.py:807]: Done

File: src/databricks/labs/ucx/contexts/application.py [hunk 7]
```diff
@@ -272,6 +277,16 @@ def table_ownership(self) -> TableOwnership:
             self.workspace_path_ownership,
         )
 
+    @cached_property
+    def static_table_ownership(self) -> StaticTableOwnership:
```
[the remark this answers is not in the document]
Comment (FastLee) [reply src/databricks/labs/ucx/contexts/application.py]: Done

File: src/databricks/labs/ucx/contexts/application.py [hunk 8]
```diff
@@ -272,6 +277,23 @@ def table_ownership(self) -> TableOwnership:
...
+                logger.warning("Default owner group is not valid, falling back to administrator ownership.")
+            else:
+                owner_group = self.config.default_owner_group
+        # Returns a static table ownership resolver
+        return DefaultSecurableOwnership(
+            self.administrator_locator,
+            self.tables_crawler,
+            owner_group,
+            lambda: self.workspace_client.current_user.me().user_name,
+        )
```
Comment (nfx) [inline src/databricks/labs/ucx/contexts/application.py:295]: this is too much logic for a factory class. reduce this method to 
```suggestion
    def default_securable_ownership(self) -> DefaultSecurableOwnership:
        return DefaultSecurableOwnership(
            self.administrator_locator,
            self.tables_crawler,
            self.group_manager,
            self.config.default_owner_group,
            self.workspace_client,
        )
```
and perform validation in `DefaultSecurableOwnership#_application_principal`

File: src/databricks/labs/ucx/hive_metastore/grants.py [hunk 9]
```diff
@@ -119,12 +119,14 @@ def object_key(self) -> str:
...
+        """Apply Ownership First, then the rest.
+        Consider Revising When we properly apply manage permission"""
+
         match self.action_type:
             case "OWN":  # Apply ownership as last to avoid losing permissions for applying grants
-                return 1
-            case _:
                 return 0
+            case _:
+                return 1
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/grants.py:129]:
```python
if self.action_type == 'OWN':
  return 0
return 1
```
no need for `match`.

File: src/databricks/labs/ucx/hive_metastore/grants.py [hunk 10]
```diff
@@ -811,18 +816,23 @@ def retrieve(self, src: SecurableObject, dst: SecurableObject) -> list[Grant]:
...
     @cached_property
     def _grants(self) -> list[Grant]:
+        # Accumulate grants from all loaders skips ownership grants
         grants = []
-        for loader in self._grant_loaders:
+        for index, loader in enumerate(self._grant_loaders):
             for grant in loader():
+                # Skip ownership grants for all loaders other than the first one
+                # The assumption is that the first one will be designated as the ownership loader
+                if index != 0 and grant.action_type == "OWN":
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/grants.py:834]: fine...

File: src/databricks/labs/ucx/workspace_access/groups.py [hunk 11]
```diff
@@ -624,6 +624,46 @@ def delete_original_workspace_groups(self):
...
+        # This method is used to find all the account groups that a user is a member of.
+        groups: list[Group] = []
+        account_groups = self._list_account_groups("id,displayName,externalId,members")
+        if not account_groups:
+            return groups
+        for group in account_groups:
+            if not group.members:
+                continue
+            for member in group.members:
+                if member.display == username:
```
Comment (nfx) [inline src/databricks/labs/ucx/workspace_access/groups.py:662]: does it work with service principal?

**Review by nfx — CHANGES_REQUESTED** [review]

File: tests/unit/hive_metastore/test_grants.py [hunk 12]
```diff
@@ -569,30 +698,58 @@ def grant_loader() -> list[Grant]:
...
+                'fake_owner',
+                'OWN',
                 catalog=catalog,
                 database=database,
                 table=table,
-            ),
+            )
         ]
 
+    def grant_loader() -> list[Grant]:
```
[the remark this answers is not in the document]
Comment (FastLee) [reply tests/unit/hive_metastore/test_grants.py:637]: Fixed

File: tests/unit/hive_metastore/test_grants.py [hunk 13]
```diff
@@ -526,36 +526,165 @@ def test_crawler_should_filter_databases() -> None:
...
+    ],
+)
+def test_migrate_grants_applies_query(
+    src: Catalog | Schema | Table,
+    grant: Grant,
+    dst: Catalog | Schema | Table,
+    query: str,
+) -> None:
+    group_manager = create_autospec(GroupManager)
+    backend = MockBackend()
```
[the remark this answers is not in the document]
Comment (FastLee) [reply tests/unit/hive_metastore/test_grants.py]: fixed

File: src/databricks/labs/ucx/contexts/application.py (hunk 8 above)
Comment (FastLee) [reply src/databricks/labs/ucx/contexts/application.py]: Fixed

File: src/databricks/labs/ucx/hive_metastore/grants.py (hunk 9 above)
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/grants.py:129]: Fixed

File: src/databricks/labs/ucx/hive_metastore/grants.py (hunk 10 above)
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/grants.py:834]: Huh

File: src/databricks/labs/ucx/workspace_access/groups.py (hunk 11 above)
Comment (FastLee) [reply src/databricks/labs/ucx/workspace_access/groups.py]: Nope! And it doesn't work with users too ;-)
Fixed

File: labs.yml [hunk 14]
```diff
@@ -358,4 +358,5 @@ commands:
   - name: enable-hms-federation
     description: (EXPERIMENTAL) Enable HMS federation based migration flow. When this is enabled, UCX will create a federated HMS catalog which syncs from the workspace HMS.
 
-
+  - name: assign-owner-group
+    description: Assign owner group to the workspace. This group will be assigned as an owner to all migrated tables and views.
```
[the remark this answers is not in the document]
Comment (FastLee) [reply labs.yml:362]: Done

File: src/databricks/labs/ucx/contexts/application.py [hunk 15]
```diff
@@ -272,6 +277,27 @@ def table_ownership(self) -> TableOwnership:
             self.workspace_path_ownership,
         )
 
+    @cached_property
+    def default_owner_group(self) -> str | None:
+        owner_group = None
+        if self.config.default_owner_group:
+            if not self.group_manager.validate_owner_group(self.config.default_owner_group):
+                logger.warning("Default owner group is not valid, falling back to administrator ownership.")
```
Comment (nfx) [inline src/databricks/labs/ucx/contexts/application.py:285]: `ApplicationContext` must only have object initialisation logic. this has to happen after constructor of `DefaultSecurableOwnership` is initialized, in one of the methods.

**`ApplicationContext` must not make calls to API.**

**Review by nfx — COMMENTED** [review]
make validation of current user to be part of a default group. also make sure people know about it in readme.

**Review by nfx — APPROVED** [review]
stamping to unblock the merge, but comments have to be addressed in a follow-up PR

File: src/databricks/labs/ucx/hive_metastore/ownership.py [hunk 16]
```diff
@@ -85,6 +87,117 @@ def _grants_snapshot(self):
...
+        table_crawler: TablesCrawler,
+        group_manager: GroupManager,
+        default_owner_group: str | None,
+        app_principal_resolver: Callable[[], str | None],
+    ) -> None:
+        self._tables_crawler = table_crawler
+        self._group_manager = group_manager
+        self._default_owner_group = default_owner_group
+        self._app_principal_resolver = app_principal_resolver
+        super().__init__(administrator_locator)
```
[the remark this answers is not in the document]
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/ownership.py:109]: Fixed

File: tests/unit/hive_metastore/test_tables.py [hunk 17]
```diff
@@ -699,10 +783,140 @@ def test_table_owner() -> None:
...
-    workspace_path_ownership.owner_of.assert_not_called()
+    assert owner == expected
+    assert admin_locator.get_workspace_administrator.called == workspace_owner
+    assert legacy_query_ownership.owner_of.called == legacy_query
+    assert workspace_path_ownership.owner_of_path.called == workspace_path
+
+
+@pytest.mark.parametrize(
+    'default_owner_group, cli_user, valid_admin, grants',
+    [
```
[the remark this answers is not in the document]
Comment (FastLee) [reply tests/unit/hive_metastore/test_tables.py:794]: Fixed

**Pull request merged**